pkg/install/installer.go

```diff
@@ -35,178 +30,127 @@ type Config struct {
-				if err := os.MkdirAll(subFolder, 0755); err != nil && !os.IsExist(err) {
-					return fmt.Errorf("failed to create directory")
+//collectArtifacts goes through the profile definition and any nested profiles to collect all artifacts
+func (i *Installer) collectArtifacts(installation profilesv1.ProfileInstallation, isNestedArtifact bool) ([]artifact.ArtifactWrapper, error) {
+	path := installation.Spec.Source.Path
+	branchOrTag := installation.Spec.Source.Tag
+	if installation.Spec.Source.Tag == "" {
+		branchOrTag = installation.Spec.Source.Branch
+	}
+	profileDef, err := i.cloneRepoAndGetProfileDefinition(installation.Spec.Source.URL, branchOrTag, path)
+	if err != nil {
+		return nil, err
+	}
+	profileRepoKey := cloneCacheKey(installation.Spec.Source.URL, branchOrTag)
+
+	var artifacts []artifact.ArtifactWrapper
+	for _, a := range profileDef.Spec.Artifacts {
+		//If its a nested profile lets make a recursive call to scans its artifacts
+		if a.Profile != nil {
+			nestedInstallation := installation.DeepCopyObject().(*profilesv1.ProfileInstallation)
+			nestedInstallation.Spec.Source.URL = a.Profile.Source.URL
+			nestedInstallation.Spec.Source.Branch = a.Profile.Source.Branch
+			nestedInstallation.Spec.Source.Tag = a.Profile.Source.Tag
+			nestedInstallation.Spec.Source.Path = a.Profile.Source.Path
+			if a.Profile.Source.Tag != "" {
+				path := "."
+				splitTag := strings.Split(a.Profile.Source.Tag, "/")
+				if len(splitTag) > 1 {
+					path = splitTag[0]
 				}
-				filename = filepath.Join(subFolder, fmt.Sprintf("%s.%s", name, "yaml"))
+				nestedInstallation.Spec.Source.Path = path
 			}
-			if err := i.generateOutput(filename, obj.Object); err != nil {
-				return err
+			nestedInstallation.Name = a.Name
+			nestedArtifacts, err := i.collectArtifacts(*nestedInstallation, true)
+			if err != nil {
+				return nil, err
 			}
-		}
-		// if we have a local resource, write out the kustomization yaml limiting its visibility.
-		if artifact.Kustomize.LocalResourceLimiter != nil {
-			// This is helmRelease related so it must be inside the sub-folder for the helm release.
-			filename := filepath.Join(artifactDir, artifact.SubFolder, "kustomization.yaml")
-			if err := writeOutKustomizeResource(artifact.Kustomize.LocalResourceLimiter, filename); err != nil {
-				return err
+			artifacts = append(artifacts, nestedArtifacts...)
+		} else {
+			newArtifact := artifact.ArtifactWrapper{
+				Artifact:           a,
+				PathToProfileClone: filepath.Join(i.clonedRepos[profileRepoKey], installation.Spec.Source.Path),
+				ProfileName:        profileDef.Name,
 			}
-		}
-		filename := filepath.Join(artifactDir, "kustomization.yaml")
-		if err := writeOutKustomizeResource(artifact.Kustomize.ObjectWrapper, filename); err != nil {
-			return err
+			//if its a nested artifact we want to wrap the artifact in a sub-directory
+			if isNestedArtifact {
+				newArtifact.NestedProfileSubDirectoryName = installation.Name
```

Comment:
This should do something like filepath.Join. Right now it doesn't nest double nested profiles. Structures like this:

```
nested artifacts A -> nested artifacts B -> nginx
```

Ends up in something like this:

```
nested artifacts from A
nested artifacts from B
```

This will cause a problem if the artifacts happened to be named the same. Then they will share a folder and overwrite each other.

Reply:
I created a test profile to test that here: https://github.com/weaveworks/profiles-examples/tree/double_nested

Reply:
mhmm good point! will update and add a test

Reply:
Added unit test for scenario, which appears to have fixed it :smile: 

Also just for sanity manaully tried our your branch, LGTM:
```
├── artifacts
│   ├── dependon-chart
│   │   ├── helm-chart
│   │   │   ├── ConfigMap.yaml
│   │   │   ├── HelmRelease.yaml
│   │   │   └── HelmRepository.yaml
│   │   ├── kustomization.yaml
│   │   └── kustomize-flux.yaml
│   ├── nested-nginx
│   │   ├── nested-profile2
│   │   │   └── nginx-server
│   │   │       ├── helm-chart
│   │   │       │   ├── HelmRelease.yaml
│   │   │       │   ├── kustomization.yaml
```

Reply:
<img width="355" alt="Screenshot 2021-07-27 at 8 38 35" src="https://user-images.githubusercontent.com/182850/127107490-a48a48ee-cd93-45e9-8076-b7ca18a89cc6.png">

I had this before your change. :) Now it does work, yes. :D

Reply:
Also, I expect there is an nginx-server besides nested profile2, right? Which is the local resource of the nested-nginx thing. :)

Reply:
<img width="345" alt="Screenshot 2021-07-27 at 8 40 18" src="https://user-images.githubusercontent.com/182850/127107683-afc2601d-5b04-42e4-a491-eb6717f31d95.png">

Basically, this is now as I would expect it. With two nested profiles each in their respective folder plus, the local resource.

Reply:
I don't quite follow, is the current behaviour correct :see_no_evil: ?